### PR TITLE
Require MySQL 8.0, PostgreSQL 12, MSSQL 2012

### LINKED ISSUE #25657: Drop MySQL 5.7 support
MySQL 5.7 will reach EOL in [4 months](https://endoflife.date/mysql). We should:

- Drop it from CI matrix and remove relevant Makefile targets
- Update docs that MySQL 8.0 or greater is required

### PR DESCRIPTION
**Pull request by silverwind**
- MySQL 5.7 support and testing is dropped
- MySQL tests now execute against 8.1, up from 5.7 and 8.0
- PostgreSQL 10 and 11 support ist dropped
- PostgreSQL tests now execute against 16, up from 15
- MSSQL 2008 support is dropped
- MSSQL tests now run against locked 2022 version

Fixes: https://github.com/go-gitea/gitea/issues/25657

Ref: https://endoflife.date/mysql
Ref: https://endoflife.date/postgresql
Ref: https://endoflife.date/mssqlserver

## :warning: BREAKING :warning:

Support for MySQL 5.7, PostgreSQL 10 and 11, and MSSQL 2008 is dropped. You are encouraged to upgrade to supported versions.

### CODE REVIEW

**Comment by silverwind** [comment]
After this, the required checks config on github needs to be updated from `test-mysql5` and `test-mysql8` to just `test-mysql`. The PR will also not go green without this change.

**Review by delvh — APPROVED** [review]

**Review by denyskon — APPROVED** [review]

**Comment by delvh** [comment]
I mean, we could also be so nice and block this PR until the mentioned versions are actually EOL.
But I'm also fine with the current approach.

**Comment by denyskon** [comment]
I don't think that people who use MySQL 5.7 also use Gitea nightly in production 😆
I think it's fine, the release will definitely happen after all of the mentioned versions are EOL.

**Comment by silverwind** [comment]
By the time the 1.22 branch releases they will be EOL. I imagine anyone running unstable builds will likely not run unsupported databases :)

**Comment by denyskon** [comment]
@silverwind ✨ same though at the same moment ✨

**Comment by delvh** [comment]
> I don't think that people who use MySQL 5.7 also use Gitea nightly in production 😆

Good point.

File: Makefile
Comment (silverwind) [inline Makefile:167]: It seems there is a discrepancy between this database name and the action which has `test` as the `MYSQL_DATABASE`. For whatever reason this has worked so far, but I'll try setting the action to `testgitea` as well.

File: .github/workflows/pull-db-tests.yml
Comment (silverwind) [inline .github/workflows/pull-db-tests.yml]:
```suggestion
          MYSQL_DATABASE: testgitea
```

File: Makefile
Comment (silverwind) [reply Makefile:167]: Maybe it's because gitea/xorm execute a `CREATE DATABASE`, so it likely wouldn't matter which database is pre-created.

Comment (silverwind) [reply Makefile:167]: Tests still pass so the change works.

**Comment by techknowlogick** [comment]
Thanks for keeping this current 😀 

Could you add a note that should someone need support for an outdated database that they can contact us for an "extended support contract" (or some other name?). As then we can then test for those older databases and add fixes into this codebase too. I know there are some people who are using the outdated mssql version that are paying MS for them to backport fixes, so they may need to have that support in Gitea too.

**Comment by silverwind** [comment]
Feel free to edit the OP to your liking :)

File: .github/workflows/pull-db-tests.yml
Comment (wxiaoguang) [inline .github/workflows/pull-db-tests.yml:89]: It should test with MySQL 8.0 but not 8.1

The same to pgsql/mssql

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml:89]: Why not test with latest? New deployments are likely to use latest. I see no point testing non-latest if we only test one version.

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml:89]: This one step I guess we could move to 8.0 because 8.1 is tested below separately, but I don't see much of a point either.

Comment (wxiaoguang) [reply .github/workflows/pull-db-tests.yml:89]: What if you used a MySQL 8.1 only syntax?

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml:89]: By your notion, we would have to test every single supported db version, which given our CI resources is likely not feasible. Also, MariaDB is completely untested.

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml:89]: Maybe we should stick to MySQL LTS versions? According to this 8.0, 8.4, 9.7 will be LTS:

image

Comment (wxiaoguang) [reply .github/workflows/pull-db-tests.yml:89]: Ideally, MariaDB should have a separate test.

For others, databases have pretty good backward compatibility. If MySQL 8.0 passes, then MySQL 8.1 are almost sure to pass.

Comment (wxiaoguang) [reply .github/workflows/pull-db-tests.yml:89]: > Maybe we should stick to MySQL LTS versions? According to this 8.0, 8.4, 9.7 will be LTS:

No, at the moment, only the minimal supported version is enough.

Comment (silverwind) [reply .github/workflows/pull-db-tests.yml:89]: I still don't understand the reasons. We can:

- Test latest, be most likely correct for new deployment, but old ones may break with incompatibly SQL syntax etc.
- Test oldest, be correct for ancient deployments, but unknown for new ones.
- Test something in between, but by what rules?

I think it's easier to maintain if we just test latest, so that the maintainer updating this doesn't have to thing about obscure rules about which version to test.

Ideally we'd have a matrix build to easily add/remove versions. Currently it's a pain because of all the Makefile changes and such. If we had a matrix, I guess testing oldest and newest of each version would be ideal.

Comment (wxiaoguang) [reply .github/workflows/pull-db-tests.yml:89]: > * Test latest, be most likely correct for new deployment, but old ones may break with incompatibly SQL syntax etc.

No, it would break the old database easily.

> * Test oldest, be correct for ancient deployments, but unknown for new ones.

This is the right approach. It won't be "unknown": see my comment: `databases have pretty good backward compatibility. If MySQL 8.0 passes, then MySQL 8.1 are almost sure to pass.`

> * Test something in between, but by what rules?

No, it's not worth to be considered.

**Comment by lunny** [comment]
Yes, if we say it required mysql 8.0, we should test the minial version 8.0 but not 8.1 like we did with Golang.

File: docs/content/installation/database-preparation.en-us.md
Comment (techknowlogick) [inline docs/content/installation/database-preparation.en-us.md:21]:
```suggestion

If you use an unsupported database version, please [get in touch](/help/support) with us for information on our Extended Support Contracts. We can provide testing and support for older databases and integrate those fixes into the Gitea codebase.

```

File: .github/workflows/pull-db-tests.yml
Comment (techknowlogick) [reply .github/workflows/pull-db-tests.yml:89]: This PR doesn't change the behaviour of what we are doing with tests (w/ psql we're using latest, and w/ mqsl we were using 8.x).

I agree we should test both min/max db versions, but that can be done in a different PR.

**Comment by wxiaoguang** [comment]
@techknowlogick 

> This PR doesn't change the behaviour of what we are doing with tests (w/ psql we're using latest, and w/ mqsl we were using 8.x).
> I agree we should test both min/max db versions, but that can be done in a different PR.

Testing the minimal version is good enough and the right approach, like before.

But I guess nobody would really follow it if this gets merged, until some bugs like #22809 occurs (pgsql 10 works but pgsql 9 doesn't -- newer version works, older version fails)

This is my last comment for this problem, because I have explained enough and I do not want to bother people.

**Comment by lunny** [comment]
ref: #27429

**Pull request merged**
